### PR TITLE
QNW controls and Rigaku fast/slow staging

### LINKED ISSUE #267: Ophyd device for QNW temperature controller
The PV name for the device is `8idi:QNWenv_1`, `8idi:QNWenv_2` and `8idi:QNWenv_3`. Tom Walsh at BCDA wrote the EPICS IOC for it. The screenshot shows the caqtdm screen at 8-ID-I. 

The device is the cornerstone for remote operation at 8-ID-I during Covid and will continue to operate after APS-U.

image

### LINKED ISSUE #270: Rigaku fast/slow mode staging not working
@prjemian  I tried to stage Rigaku detector to acquire in fast mode, however the behavior (screenshot) is quite different from [what we had in March](https://github.com/aps-8id-dys/ipython-8idiuser/issues/251#issuecomment-790309503) and the parameters for the ZDT mode is not passing into the IOC.

Any suggestions?

image

### PR DESCRIPTION
**Pull request by qzhang234**
Fix #267 (work-in-progress) and fix #270

### CODE REVIEW

File: profile_bluesky/startup/instrument/devices/ad_rigaku_detector.py [hunk 1]
```diff
@@ -1,9 +1,7 @@
 """
 ADRigaku UFXC area detector (EPICS)
-
```
Comment (prjemian) [inline profile_bluesky/startup/instrument/devices/ad_rigaku_detector.py:3]: please leave this blank line

File: profile_bluesky/startup/instrument/devices/ad_rigaku_detector.py [hunk 2]
```diff
@@ -1,9 +1,7 @@
...
 * detector name: -tba-
 * detector number: -tba-
-
```
Comment (prjemian) [inline profile_bluesky/startup/instrument/devices/ad_rigaku_detector.py:6]: This blank line should stay also (it is part of the restructured text formatting)

File: profile_bluesky/startup/instrument/devices/ad_rigaku_detector.py [hunk 3]
```diff
@@ -39,7 +37,6 @@
 class RigakuUfxcDetectorCam(AD_AcquireDetectorCamBase, CamBase):
     """
     Customization for the additional fields of the ADRigaku detector.
-
```
Comment (prjemian) [inline profile_bluesky/startup/instrument/devices/ad_rigaku_detector.py:42]: leave this blank line in place (and similar for other places)

File: profile_bluesky/startup/instrument/devices/ad_rigaku_detector.py [hunk 4]
```diff
@@ -63,22 +60,6 @@ class RigakuUfxcDetectorCam(AD_AcquireDetectorCamBase, CamBase):
     pool_max_buffers = None
     EXT_TRIGGER = 0
 
-    def setup_modes(self, num_triggers):
```
Comment (prjemian) [inline profile_bluesky/startup/instrument/devices/ad_rigaku_detector.py:66]: This method is _required_ for the `AD_Acquire()` plan to work.  If it is deleted, the base class will raise a `NotImplementedError()` exception.

Comment (prjemian) [reply profile_bluesky/startup/instrument/devices/ad_rigaku_detector.py:66]: If there is no work to be done (from bluesky's point of view), then:

```python
        yield from bps.null()  # at least must yield *some* bluesky message
```

Comment (prjemian) [reply profile_bluesky/startup/instrument/devices/ad_rigaku_detector.py:66]: Ah, there it is below.

File: profile_bluesky/startup/instrument/devices/ad_rigaku_detector.py [hunk 5]
```diff
@@ -131,8 +118,7 @@ def staging_setup_DM(self, *args, mode=None):
         # unix(sprintf("chmod 777 %s", UFXC_fullpath_datafolder))
         print("DEBUG: file_path (bluesky): ", file_path)
         print("DEBUG: file_path (detector): ", fpath)
-        os.makedirs(file_path, mode=0o777)
-        os.chmod(file_path, 0o777)
+        os.makedirs(file_path, mode=0o777,exist_ok=True)
```
Comment (prjemian) [inline profile_bluesky/startup/instrument/devices/ad_rigaku_detector.py:121]: Good.  This creates intermediate directories as needed.

File: profile_bluesky/startup/instrument/devices/ad_rigaku_detector.py [hunk 6]
```diff
@@ -171,14 +162,10 @@ def staging_setup_DM(self, *args, mode=None):
             self.stage_sigs["imm1.file_name"] = "test"
             # TODO: what else is needed?
```
Comment (prjemian) [inline profile_bluesky/startup/instrument/devices/ad_rigaku_detector.py:163]: now is time to remove this TODO line

File: profile_bluesky/startup/instrument/devices/ad_rigaku_detector.py [hunk 7]
```diff
@@ -171,14 +162,10 @@ def staging_setup_DM(self, *args, mode=None):
...
-        self.stage_sigs["cam.file_name"] = os.path.basename(f"{fname}.bin")
-
     @property
     def images_received(self):
         """
         Return the number (int) of images captured.
-
         suggestion:  ``self.immout.num_captured.get()``
         """
         raise NotImplementedError("Must implement in detector-specific subclass.")
```
Comment (prjemian) [inline profile_bluesky/startup/instrument/devices/ad_rigaku_detector.py:171]: Is there work to do here?  Won't this exception be a problem when we try to use it?

File: profile_bluesky/startup/instrument/devices/ad_rigaku_detector.py [hunk 8]
```diff
@@ -225,4 +209,4 @@ def xpcs_loop(self, *args, **kwargs):
 logger.info("Sleeping for %s seconds before creating adrigaku object.", _delay)
 ttime.sleep(_delay)
 
-adrigaku = RigakuUfxcDetector(IOC_PREFIX, name="adrigaku")
+adrigaku = RigakuUfxcDetector(IOC_PREFIX, name="adrigaku")
```
Comment (prjemian) [inline profile_bluesky/startup/instrument/devices/ad_rigaku_detector.py:212]: add a blank line at the end of _every_ Python file

- see: https://stackoverflow.com/a/2287990/1046449
- Following these guidelines will make code easier to review: [PEP 8 -- Style Guide for Python Code](https://www.python.org/dev/peps/pep-0008/)

File: profile_bluesky/startup/instrument/devices/lambda_750k.py [hunk 9]
```diff
@@ -355,21 +355,21 @@ def getCounts(self):
         # pseudo-counter named "ccdc" (that's what chk_ccdc() does)
         if self.chk_ccdc:
             if dm_pars.compression.get() == 1:
-                return self.immout.num_pixels.get()
+                return self.imm1.num_pixels.get()
```
Comment (prjemian) [inline profile_bluesky/startup/instrument/devices/lambda_750k.py:358]: will be part of #274

Comment (prjemian) [reply profile_bluesky/startup/instrument/devices/lambda_750k.py:358]: and below, as well

Comment (qzhang234) [reply profile_bluesky/startup/instrument/devices/lambda_750k.py:358]: Got it. Should I merge now and continue to work on it via #274 later?

File: profile_bluesky/startup/instrument/devices/util_components.py [hunk 10]
```diff
@@ -0,0 +1,155 @@
...
+    def __init__(self, *args, readback_attr='readback',
+                 setpoint_attr='setpoint', tolerance_attr='tolerance',
+                 **kwargs):
+        super().__init__(*args, **kwargs)
+        self._readback_attr = readback_attr
+        self._setpoint_attr = setpoint_attr
+        self._tolerance_attr = tolerance_attr
+
+    def get(self, **kwargs):
+        readback = getattr(self.parent, self._readback_attr)
```
Comment (prjemian) [inline profile_bluesky/startup/instrument/devices/util_components.py:25]: @qzhang234 , @gfabbris : Are the attribute names expected to change after an instance of this class is created?  If not, then the `getattr()` can be moved into the `__init__()` so they are done only once.

Comment (gfabbris) [reply profile_bluesky/startup/instrument/devices/util_components.py:25]: Yes, all these `getattr()` should be in the `__init__()`. But these two utils (`DoneSignal` and `TrackingSignal`) is outdated (I will remove from the 4-ID-D setup), and I think it's not used anywhere here, so these could just be removed.

File: profile_bluesky/startup/instrument/devices/ad_imm_plugins.py [hunk 11]
```diff
@@ -19,91 +18,91 @@
...
-        for plugin in (self.imm0, self.imm1, self.imm2, self.immout):
-            if plugin.file_format.get() not in (0, "IMM_Raw"):
-                yield from bps.mv(
-                    plugin.capture,
-                    "Done",  # ('Done', 'Capture')
-                    plugin.file_format,
-                    "IMM_Raw",  # ('IMM_Raw', 'IMM_Cmprs')
-                )
+    # imm1 = Component(IMMnLocal, "IMM1:")
+    imm1 = Component(IMMoutLocal, "IMM1:")
```
Comment (prjemian) [inline profile_bluesky/startup/instrument/devices/ad_imm_plugins.py:75]: Is this in use now?

**Pull request merged**